corehq/apps/smsbillables/models.py

```diff
@@ -27,23 +28,43 @@ class SmsGatewayFeeCriteria(models.Model):
     backend_instance = models.CharField(max_length=255, db_index=True, null=True)
     direction = models.CharField(max_length=10, db_index=True, choices=DIRECTION_CHOICES)
     country_code = models.IntegerField(max_length=5, null=True, blank=True, db_index=True)
+    prefix = models.CharField(max_length=10, blank=True, default="", db_index=True)
 
     @classmethod
-    def get_most_specific(cls, backend_api_id, direction, backend_instance=None, country_code=None):
+    def get_most_specific(cls, backend_api_id, direction,
+                          backend_instance=None, country_code=None, national_number=""):
         """
         Gets the most specific criteria available based on (and in order of preference for optional):
         - backend_api_id
         - direction
         - backend_instance (optional)
-        - country_code (optional)
+        - country_code and prefix (optional)
         """
         all_possible_criteria = cls.objects.filter(backend_api_id=backend_api_id, direction=direction)
 
         if all_possible_criteria.count() == 0:
             return None
 
+        def get_criteria_with_longest_matching_prefix(criteria_list):
+            if len(set(criteria.prefix for criteria in criteria_list)) != len(criteria_list):
+                raise AmbiguousPrefixException(
+                    ", ".join(
+                        "%(country_code)d, '%(prefix)s'" % {
+                            "country_code": criteria.country_code,
+                            "prefix": criteria.prefix,
+                        } for criteria in criteria_list
+                    )
+                )
+            criteria_list.sort(key=(lambda criteria: len(criteria.prefix)), reverse=True)
+            for criteria in criteria_list:
+                if national_number.startswith(criteria.prefix):
```

Comment:
Now that national_number can be empty string, should we also check here for:

    if criteria.prefix and ...

because otherwise empty string can match empty string